src/procedural-macros.md

```diff
@@ -1,24 +1,546 @@
+    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
+     Running `target/debug/foo`
+the answer was: 4
+```
+
+showing us how the input was ignored and the macro's output was used instead.
+
+### Custom Derive
+
+[The book][procedural macros] has a tutorial on creating custom derives and here
+we'll go into some of the nitty-gritty of how this works. The custom derive
+feature allows you to define a custom `#[derive(Foo)]` mode which often makes it
+much easier to systematically implement traits, removing quite a lot of
+boilerplate.
+
+Custom derives are defined like so:
+
+```rust,ignore
+#[proc_macro_derive(MyTrait)]
+pub fn foo(item: TokenStream) -> TokenStream {
+    // ...
+}
+```
+
+Here the argument to the `proc_macro_derive` attribute, `MyTrait`, is the name
+of the custom derive mode. The name of the function here, `foo`, is not
+currently used (but it may one day be used).
+
+Like procedural bang macros the input to the macro here is the item that the
+attribute was applied to. Unlike bang macros, however, the output is *appended*
+to the program rather than replacing the item its attached to. We can see this
+behavior by defining a macro like:
+
+```rust,ignore
+extern crate proc_macro;
+use proc_macro::*;
+
+#[proc_macro_derive(MyTrait)]
+pub fn foo(item: TokenStream) -> TokenStream {
+    println!("{:#?}", item);
+    "fn answer() -> u32 { 2 }".parse().unwrap()
+}
+```
+
+using it liek:
+
+```rust,ignore
+extern crate my_macro;
+
+use my_macro::MyTrait;
+
+#[derive(MyTrait)]
+struct Foo;
+
+fn main() {
+    drop(Foo);
+    println!("the answer was: {}", answer());
+}
+```
+
+and compiling it:
+
+```sh
+$ cargo run
+   Compiling my-macro v0.1.0 (file://.../foo/my-macro)
+   Compiling foo v0.1.0 (file://.../foo)
+TokenStream [
+    Ident { ident: "struct", span: #0 bytes(67..73) },
+    Ident { ident: "Foo", span: #0 bytes(74..77) },
+    Punct { ch: ';', spacing: Alone, span: #0 bytes(77..78) }
+]
+    Finished dev [unoptimized + debuginfo] target(s) in 0.34s
+     Running `target/debug/foo`
+the answer was: 2
+```
+
+Here we can see how the input to the macro was a `TokenStream` representing
+the three input tokens `struct Foo;`. While our output only contained the
+`answer` function, we were still able to use `Foo` in the main program because
+custom derives *append* items, they don't replace them.
+
+Now this is a pretty wonky custom derive, and would likely be confusing to
+users! Custom derives are primarily geared towards implementing traits, like
+`Serialize` and `Deserialize`. The `syn` crate also has a [number of
+examples][synex] of defining custom derives.
+
+[procedural macros]: ../book/first-edition/procedural-macros.html
+[synex]: https://github.com/dtolnay/syn/tree/master/examples
+
+#### Custom attributes on Custom derive
+
+An additional feature of custom derive macros is that they can whitelist names
+of attributes which are considered not part of normal attribute macro expansion.
```

Comment:
We should pick a clear name for such attributes to distinguish them from attributes that are procedural macros. This is one of the most confusing aspects of the procedural macro API (practically everybody believes `#[serde(...)]` is itself a macro).

I have been using the terms "attribute macro" and "inert attribute" to distinguish the two.

Reply:
I recently had to come up with a name for this in the compiler and used "derive helper attribute".

"Inert attribute", if it means non-macro attribute, an attribute that doesn't expand the item it's applied to, can currently refer to a built-in attribute (`#[inline]`), a tool attribute (`#[rustfmt::skip]`), a "derive helper" (`#[serde(...)]`), or a "custom" attribute on nightly (`feature(custom_attribute)` + `#[my_attr]`).